x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,405 @@
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			// read from s3 object and create event for each log line
+			p.readS3CreateEvents(svcS3, s3Infos)
+
+			// delete message after events are sent
+			err = deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+			}
+		}(messages[i])
+
+		go func(message sqs.Message) {
+			for {
+				select {
+				case <-p.close:
+					return
+				case <-done:
+					return
+				case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
+					// if half of the set visibilityTimeout passed and this is
+					// still ongoing, then change visibility timeout.
+					changeMessageVisibilityInput.ReceiptHandle = message.ReceiptHandle
+					req := svcSQS.ChangeMessageVisibilityRequest(changeMessageVisibilityInput)
+					_, err := req.Send()
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+					}
+				}
+			}
+		}(messages[i])
+	}
+	wg.Wait()
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleMessage(m sqs.Message) ([]s3Info, error) {
+	msg := map[string]interface{}{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	records := msg["Records"].([]interface{})
+	for _, record := range records {
+		recordMap := record.(map[string]interface{})
+		if recordMap["eventSource"] == "aws:s3" && recordMap["eventName"] == "ObjectCreated:Put" {
+			s3Info := s3Info{}
+			s3Info.region = recordMap["awsRegion"].(string)
+			s3Record := recordMap["s3"].(map[string]interface{})
+
+			bucketInfo := s3Record["bucket"].(map[string]interface{})
+			s3Info.name = bucketInfo["name"].(string)
+
+			objectInfo := s3Record["object"].(map[string]interface{})
+			s3Info.key = objectInfo["key"].(string)
+			s3Infos = append(s3Infos, s3Info)
+		}
+	}
+	return s3Infos, nil
+}
+
+// stringInSlice checks if a string is already exists in list
+// If there is no bucketNames configured, then collect all.
+func stringInSlice(name string, bucketNames []string) bool {
+	if bucketNames == nil || len(bucketNames) == 0 {
+		return true
+	}
+
+	for _, v := range bucketNames {
+		if v == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Input) readS3CreateEvents(svc s3iface.S3API, s3Infos []s3Info) {
+	if len(s3Infos) > 0 {
+		var wg sync.WaitGroup
+		numS3Infos := len(s3Infos)
+		wg.Add(numS3Infos)
+
+		for i := range s3Infos {
```

Comment:
Currently, all S3 event notifications have a single event per notification message. So it will be only one `s3Infos` per message. But AWS said they might put more events into one message later, that's why I set `s3Infos` as an array.

Reply:
I am also concerned about the number of goroutines we are producing in this input, I would really like a bounded approach to control the concurrency.

Because right now we have:

```
Number of SQS queues * Number of messages (batch default is 10 IIRC) * 2 * Number of S3Info
```

Since everything is async it means that on each call to `Run()` could potentially just increase the number of goroutines used, which make me investigate how back pressure is handled in this input, lets say that the output is blocked for an indefinite period of time, I think the current implementation of the S3 input will just generate an unbounded quantity of goroutines until the process ran out of memory? Is that correct @kaiyan-sheng ?

Reply:
Yeah good point, I think making this a bounded approach is needed. 
Maximum number of messages will be 10 per each collection period. Each message usually points to one s3 object but that might change in the future if AWS decides to put more events into one message.
@ph Is it ok for me to create a new PR to solve this issue?

Reply:
I will continue working on this PR till this problem is solved 😬 not safe to merge it with potential memory issue!

Reply:
With the many number of potential go-routines I wonder if it would make sense to solve this using a counting semaphore. On init we configure the semaphore with the maximum number of active go-routines.